### PR TITLE
Update `tests/README.md` to document recent changes to integration testing strategy (A/B-testing, new fixtures, etc.)

### PR DESCRIPTION
**Pull request by roypat**
## Changes

- Documents our A/B-Testing strategy, including a brief description of what A/B-testing is, how and when we use it, what it means to execute an A/B-Test on a PR vs executing it outside of a PR (e.g. inside a nightly pipeline), how post-merge performance A/B-tests differ from pre-merge A/B-tests
- Document how to parameterize a test by all supported guest kernels by mentioning the key differences between `uvm_plain` and `uvm_plain_any` fixtures
- Document the pytest markers `nonci` and `no_block_pr` and explain which tests are run when
- Various small documentation fixups
- Eliminate usage of `test_microvm_with_api` and replace it with `uvm_plain`/`uvm_plain_any` as appropriate

## Reason

Many things weren't documented yet

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following Developer
Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [ ] If a specific issue led to this PR, this PR closes the issue.
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this
  PR.
- [ ] API changes follow the [Runbook for Firecracker API changes][2].
- [ ] User-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.
- [ ] New `TODO`s link to an issue.
- [ ] Commits meet
  [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

______________________________________________________________________

- [ ] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: tools/ab_test.py [hunk 1]
```diff
@@ -136,7 +136,9 @@ def load_data_series(report_path: Path, revision: str = None, *, reemit: bool =
                 else:
                     # If there are many data points for a metric, they will be split across
                     # multiple EMF log messages. We need to reassemble :(
-                    assert processed_emf[dimension_set].keys() == result.keys()
+                    assert (
```
Comment (kalyazin) [inline tools/ab_test.py:139]: commit message `Improve message if 1:1 mapping of metrics<->dimensions is broken`: can add `test` tag.

File: tests/integration_tests/functional/test_balloon.py [hunk 2]
```diff
@@ -205,7 +202,7 @@ def test_deflate_on_oom(uvm_plain_any, deflate_on_oom):
 
     # We get an initial reading of the RSS, then calculate the amount
     # we need to inflate the balloon with by subtracting it from the
-    # VM size and adding an offset of 10 MiB in order to make sure we
+    # VM size and adding an offset of 20 MiB in order to make sure we
```
Comment (kalyazin) [inline tests/integration_tests/functional/test_balloon.py:205]: looks like the offset is still 10M

File: tests/integration_tests/functional/test_balloon.py [hunk 3]
```diff
@@ -185,7 +185,11 @@ def test_deflate_on_oom(test_microvm_with_api, deflate_on_oom):
 
       should result in balloon_stats['actual_mib'] remain the same
     """
-    test_microvm = test_microvm_with_api
+    if uvm_plain_any.guest_kernel_version < (5, 10, 0) and deflate_on_oom:
+        # might just be because 4.14 guest needs less memory than 5.10
+        pytest.skip("FIXME: doesn't work")
```
[the remark this answers is not in the document]
Comment (roypat) [reply tests/integration_tests/functional/test_balloon.py]: Ah, its fixed in a follow up commit, I'll squash them together

File: tests/README.md [hunk 4]
```diff
@@ -5,56 +5,68 @@ contracts of Firecracker.
...
 ```sh
-tools/devtool test -- <test_dir_or_file_path>...
+./tools/devtool test -- integration_tests/performance/test_boottime.py
 ```
 
 To run a single specific test from a file:
 
 ```sh
-tools/devtool test -- <test_file_path>::<test_name>
+./tools/devtool -y test -- integration_tests/performance/test_boottime.py::test_boottime
```
[the remark this answers is not in the document]
Comment (roypat) [reply tests/README.md:29]: no reason, just me forgetting to put it. It should be used everywhere :)

File: tests/integration_tests/performance/test_boottime.py [hunk 5]
```diff
@@ -167,7 +167,7 @@ def find_events(log_data):
 )
 @pytest.mark.nonci
 def test_boottime(
-    microvm_factory, guest_kernel, rootfs, vcpu_count, mem_size_mib, metrics
+    microvm_factory, guest_kernel, rootfs_rw, vcpu_count, mem_size_mib, metrics
```
[the remark this answers is not in the document]
Comment (roypat) [reply tests/integration_tests/performance/test_boottime.py:169]: Added a bullet point about the fixture commit

File: tests/README.md [hunk 6]
```diff
@@ -107,6 +107,132 @@ Unlike unit tests, Rust integration tests are each run in a separate process.
...
+depends on a comparison of these outputs of the test function in these two
+environment. The advantage of A/B-testing is that it does not require the
+specification of a ground truth to compare against. It is instead dynamically
+generated by running the test function in environment _A_. Firecracker's
+A/B-testing generally compares Firecracker binaries compiled from two separate
+commits (e.g. an _A_ binary which is compiled from the HEAD of the main branch,
+and a _B_ binary which is compiled from the HEAD of a pull request opened
+against main).
+
+We use this testing approach is a test's ground truth...
```
[the remark this answers is not in the document]
Comment (roypat) [reply tests/README.md:126]: actually, it should be `if` xP

File: tests/integration_tests/functional/test_balloon.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (roypat) [reply tests/integration_tests/functional/test_balloon.py]: Actually, I'll just remove this check from the fixture commit and reorder the fix after - would like to keep this commit as insubstantial as possible

File: tests/integration_tests/functional/test_balloon.py (hunk 2 above)
Comment (roypat) [reply tests/integration_tests/functional/test_balloon.py]: Ah, yeah, fixed. Originally my approach for fixing it was increasing this to 20, forgot to revert the comment change :)

File: tests/README.md [hunk 7]
```diff
@@ -107,6 +107,132 @@ Unlike unit tests, Rust integration tests are each run in a separate process.
 To learn more about Rust integration test, see
 [the Rust book](https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests).
 
+## A/B-Tests
+
+A/B-Testing is a testing strategy where some test function is executed twice in
+different environment (the A and B environment), and the overall test result
```
[the remark this answers is not in the document]
Comment (roypat) [reply tests/README.md:116]: Ah, yeah, I think it is. Fixed :)

File: tests/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (roypat) [reply tests/README.md:126]: fixed

File: tests/README.md [hunk 8]
```diff
@@ -107,6 +107,132 @@ Unlike unit tests, Rust integration tests are each run in a separate process.
...
+the script considers significant.
+
+Writing your own A/B-Test is easy: Simply write a test that outputs a data
+series and has no functional assertions. Then, when this test is run under the
+A/B-Test orchestrator, all data series emitted will be picked up automatically
+for statistical analysis.
+
+To add a new A/B-Test to our post-PR test suite, add the corresponding test
+function to [`.buildkite/pipeline_perf.py`](../.buildkite/pipeline_perf.py). To
+manually run an A/B-Test, either use
```
[the remark this answers is not in the document]
Comment (roypat) [reply tests/README.md]: Ah yeah, this shouldn't have an `either`. Fixed

File: tools/ab_test.py (hunk 1 above)
Comment (roypat) [reply tools/ab_test.py:139]: The tag would push the commit title beyond 70 characters

Comment (kalyazin) [reply tools/ab_test.py:139]: discussed offline, it's at author's discretion

**Review by kalyazin — DISMISSED** [review]

File: tests/README.md [hunk 9]
```diff
@@ -5,58 +5,73 @@ contracts of Firecracker.
...
 
-This will download test microvm images from the default test resource S3 bucket
-and run all available tests.
-
-To run tests from specific directories and/or files:
+To run only tests from specific directories and/or files:
 
 ```sh
-tools/devtool test -- <test_dir_or_file_path>...
+./tools/devtool -y test -- integration_tests/performance/test_boottime.py
```
Comment (zulinx86) [inline tests/README.md:23]: Can we unify the devtool commands into either `./tools/devtool` or `tools/devtool` in this doc?

Comment (roypat) [reply tests/README.md:23]: oh, yeah, sure :)

Comment (roypat) [reply tests/README.md]: Done, it should all just be `tools/devtool` now

**Review by kalyazin — APPROVED** [review]

**Review by pb8o — APPROVED** [review]

**Pull request merged**